src/main/java/frc/robot/utils/MovingAverage.java

```diff
@@ -0,0 +1,114 @@
+    }
+
+    public MovingAverage(Map<Double, Double> map) {
+        this.distanceVelocityMap = new TreeMap<>(map); // just because I want to order the values by the key.
+    }
+
+    public MovingAverage(double[][] values) {
+        this.distanceVelocityMap = new TreeMap<>();
+        for (double[] pair : values) {
+            if (pair.length >= 2)
+                this.distanceVelocityMap.put(pair[0], pair[1]);
+        }
+    }
+
+    public MovingAverage(Reader reader) {
+        this.distanceVelocityMap = new TreeMap<>();
+        read(reader);
+    }
+
+    private void read(Reader reader) {
+        String line = null;
+        try (BufferedReader bufferedReader = new BufferedReader(reader)) {
+            while ((line = bufferedReader.readLine()) != null) {
+                String[] pair = line.split(DELIMITER);
+                this.distanceVelocityMap.put(Double.parseDouble(pair[0]), Double.parseDouble(pair[1]));
+            }
+        } catch (IOException e) {
+            System.err.println("MovingAverage class can't read the file");
+            System.out.println(e.getMessage());
+        } catch (IndexOutOfBoundsException e) {
+            System.err.println("Please check that every line has a pair of values, the problem occurred on this line " + line);
+        }
+    }
+
+    /**
+     * Retrieve an approximation for the velocity based on the distance.
+     *
+     * @param distance the distance from the target.
+     * @return an approximation of the velocity to apply
+     */
+    public double estimateVelocity(double distance) {
+        double[][] nearest = getClosestDistance(distance);
+
+        if (nearest[0][0] == nearest[1][0]) // we have already have a record of the distance, just return the velocity based of that
+            return nearest[0][1];
+
+        // y = mx + b
+        // m = delta y / delta x
+        double m = (nearest[1][1] - nearest[0][1]) / (nearest[1][0] - nearest[0][0]);
+        // b = y - m*x
+        double b = nearest[1][1] - m * nearest[1][0];
+
+        return distance * m + b;
+    }
+
+    /**
+     * get a 2 pairs of (distance, velocity) of the closets distances below and above.
+     *
+     * @param distance the distance to return the pairs.
+     * @return a 2d array represents the closest values(below {distance, velocity}, above {distance, velocity}).
+     */
+    private double[][] getClosestDistance(double distance) {
+        double[][] closest = {{0, 0}, {Double.MAX_VALUE, 0}}; //{min, max}
+        this.distanceVelocityMap.forEach((key, value) -> {
+            double currentDistance = key, velocity = value; //we need to do it because comparison of the Double wrapper-class is off
+            double difference = currentDistance - distance;
+            if (difference == 0) {
+                closest[0][0] = currentDistance;
+                closest[0][1] = velocity;
+                closest[1][0] = currentDistance;
+                closest[1][1] = velocity;
```

Comment:
```suggestion
                closest[1] = closest[0];
```
I think it's more clear that way you want to return the same value.